### DISCUSSION TITLE
Pitfalls, honest draw backs of alpine.js

### DISCUSSION THREAD

**Discussion by torian257x** [alpinejs/alpine#749] [4. General]
Hi guys

Love the idea of alpine.js! The way react / vue work seem to be more niche honestly, at least for me personally.

Can someone who spent time with alpine.js be honest about the current state and pitfalls?

Like, learn yet another thing, to discover x,y,z all are highly problematic.

I would be very thankful if someone can be upfront of the current pitfalls and problems. For one I saw there are performance problems. To what extent is that? is it a show stopper? where is the danger line?

What else have you guys discovered?

**Comment by ryangjchandler** [comment]
@Jossnaz - thanks for opening this discussion, a great place to talk about this sort of stuff.

You won't notice a huge performance impact using Alpine unless you're using it to hold state over a massive DOM tree. For example, using libraries that create graphs and hundreds of DOM nodes will, of course, have a performance impact, especially with Alpine since it has to walk over an entire element tree. My best tip is to keep your components as small as possible. Never put `x-data` on the `<body>` or `<html>` tag.

Some people don't like the idea of Alpine since it sits directly in your markup. If you use a server-side framework that has a somewhat decent templating engine, then this can quickly be forgotten about since you can hide it inside of partial templates.

For the 99% of use-cases, Alpine is pretty fast and easy to work with. If you need to toggle some classes depending on a piece of state, Alpine will be much closer to vanilla JS in terms of speed compared to Vue or React.

The current state of Alpine is great. The community and ecosystem are growing, there are a number of third-party packages for [helper functions](https://github.com/alpine-collective/magic-helpers), [state management](https://github.com/ryangjchandler/spruce), [turbolinks integrations](https://github.com/SimoTod/alpine-turbolinks-adapter), the list goes on.

The library is also at it's most stable. Lots of small annoyances have been fixed and newer features like `x-spread` can give you some of the reusability that Vue and React offer.

Honestly, I don't think there is a danger line with Alpine for those common tasks that you probably found yourself using jQuery for back in the day. If you're getting into the realm of templating on the front-end, Alpine can do that with `x-for` and `x-if`, but the complex tasks on that side of things are better suited to Vue, React, Svelte, etc.

If you're trying to build an SPA, Alpine is definitely not the right answer. If you have server-rendered content and don't want to live in 2008 with jQuery, or spend years writing vanilla JavaScript to toggle some classes and show a `<div>`, then Alpine is definitely worth taking for a spin.

There is also a strong support group on [Discord](https://discord.com/invite/snmCYk3), the GitHub repo is always being looked at (and after) by people, including myself. As time goes on, these things will only get stronger.

The final thing I'd like to say is that every single framework that has ever existed has had growing pains and teething problems. As time goes on, they get stronger and if you find there is something drastically wrong with the framework, I'm 99% sure we can find a solution and help you out.

I'd love to know what you think of the library too. Feel free to comment here with some of your thoughts, or message me on [Twitter](https://twitter.com/ryangjchandler).

Have a good one!

**Reply by torian257x** [reply]
I highly appreciate your comment.

the problem I seem to have is that you would use jquery quite differently than alpine.  Toggle a class ok, you'd do with jquery, but I doubt alpine is the right choice to toggle a class honestly. 

Alpine is more powerful so everyone will use its functionalities a lot more. Which will lead to performance problems I guess..

**Reply by ryangjchandler** [reply]
No, I think Alpine is perfect for toggling a class. It's much lighter than jQuery, you don't need to write any `<script>` tags, everything sits on the markup. That's my opinion of course. 

I've honestly pulled Alpine into projects just to do a simple show and hide of a menu or drop down.

**Reply by ryangjchandler** [reply]
It's also worth noting that Alpine doesn't use a virtual DOM, it's all raw DOM operations so the performance impact would be unnoticeable, or negligible. If anything, performance problems are going to come from poorly optimised code in user land, that includes markup and JavaScript expressions.

**Reply by torian257x** [reply]
the thing is, you don't want to optimize code. You want to write code without having to think of optimizing it.
afaik vue is doing very well there, react is limping behind. The main difference is that vue takes care of dependency and what needs updating while in react, one has to do that manually. And if you forget (in react), it will lead to VERY hard to reproduce bugs.

so if you ever go in the direction of improving performance for updating, then make sure you do it the vue way, not the react way.

**Comment by rhengles** [comment]
Well, I actually built an app using `x-data` on the `<html>` tag 😅

It is not a very large app, but it is medium sized. It runs sufficiently well, except for one thing -
as per issue #286, the problem is I use a lot of `x-html` to print my app's translation, and the problem is that every unrelated change triggers an update on all of the `x-html`, which causes the content to blink.

I think Alpine could be very performant and have no trouble being used in the root element if it were to keep track of which elements depended on which variables, and only update those that needed it. In fact, I'm using snabbdom after alpine to diff the dom, and the performance is quite satisfactory to me.

**Reply by ryangjchandler** [reply]
Yeah, if the content you're inserting with `x-html` is large enough, you'd have a flicker without Alpine and using `el.innerHTML` directly.

**Reply by ryangjchandler** [reply]
I should've added that Alpine can definitely be improved here. In my opinion, you should avoid x-html as much as you can, since it's vulnerable to XSS and, as you mentioned, updated and re-set on each render / state change. 

To fix this though, Alpine would need to implement dependency tracking or diff the DOM like you are. Perhaps v3 will have something to fix this 👀

**Comment by romanhuba** [comment]
Hey! I’m curious to know what has changed since then. What still holds up today, and what doesn’t? How do you see Alpine.js differently compared to four years ago?

**Reply by Jipok** [reply]
I don't know how it was before, because I started using Alpine a year ago. But I see people often say that it is for small projects and that it is better to take Vue/Recat for SPA. Well, I made a pretty fat SPA and overall I am happy with everything. In some places there is a slowdown, but that is because I did not care about performance at all. For example, I have 200+ tabs and each has a very complex set of buttons. Which is completely duplicated with `x-for` for each tab. There is also `alpine-auto-animate` and a lot of logic. And only recently, having switched to a very weak PC, I noticed that the site "rendered" for a couple of seconds when opened. This is quite easily fixed. And that's all I noticed in terms of performance.

**Reply by Jipok** [reply]
The library is certainly not perfect. And from time to time strange behavior and [bugs](https://github.com/alpinejs/alpine/discussions/4552) occur. But these are very rare cases and can be solved with crutches.

**Reply by Jipok** [reply]
I also can't help but note that the idea of ​​server rendering is being promoted everywhere, using htmx, templ, livewire and other things. I used htmx+templ for some time and tried in [datastar](https://data-star.dev/). And I didn't like it. It was somehow inconvenient and stressful to code all the time. In the end, I switched to SPA with Alpine. The page is given once and then just communication with the server via JSON API. I liked it, I'm making my own "microframework" to simplify development.

**Reply by vseryakov** [reply]
hi, i am with you on this, building SPA with Alpine and happy with it, decided to opensource the very thin layer, check it out https://github.com/vseryakov/alpinejs-app

**Reply by Jipok** [reply]
This is interesting, but not at all how I use it. I wasn't going to post my work anytime soon, but you motivated me. I removed everything experimental and unstable and made a todo example: https://github.com/Jipok/JAlpine
I hated the web for too long. And although I am forced to write a frontend (and have come to love alpine ), I still hate npm, node, and all that stuff. That's why I don't use any of that. The code is given to the browser almost as is, in "templates" only include of other files is supported, and setting initial values ​​so that the page doesn't immediately pull some data from the server after loading.
Although I called it SPA. The state is essentially managed from the server. For example, check out how after creating a new record, the input field is cleared by the server side.

**Reply by Jipok** [reply]
Despite the fact that the HATEOAS principle is completely violated here, the development itself somehow reminds me of htmx. Because there is almost no code on the client side and the state is also managed from the server. But you don’t need to mix html with go code and recompile the code with the slightest change in ui.

**Reply by vseryakov** [reply]
cool stuff!
